src/parser/core/modules/AoEUsages.tsx

```diff
@@ -0,0 +1,139 @@
+import {t} from '@lingui/macro'
+import {Plural, Trans} from '@lingui/react'
+import {ActionLink} from 'components/ui/DbLink'
+import {Action} from 'data/ACTIONS/ACTIONS'
+import Module, {dependency} from 'parser/core/Module'
+import {AoeEvent} from 'parser/core/modules/AoE'
+import {SeverityTiers} from 'parser/core/modules/BuffWindow'
+import Suggestions, {SEVERITY, TieredSuggestion} from 'parser/core/modules/Suggestions'
+import React from 'react'
+import {Table} from 'semantic-ui-react'
+
+export interface AoeAbility {
+	/**
+	 * The AoE ability to check usages of.
+	 */
+	aoeAbility: Action
+	/**
+	 * The single target abilities that should be used in place of the AoE
+	 * ability when there are not enough targets present.
+	 */
+	stAbilities: Action[]
+	/**
+	 * The minimum number of targets the AoE ability should hit per cast.
+	 */
+	minTargets: number
+}
+
+/**
+ * This module checks for usages of AoE skills that have single target alternatives,
+ * such as BLMs Xenoglossy or NINs Hellfrog Medium, to ensure they hit the minimum
+ * number of targets.  It should not be used to check skills that can hit AoE but do
+ * not have alternatives, such as DRGs Dragonfire Dive or SMNs Deathflare, as it is
+ * correct to use those skills on single targets.
+ *
+ * For the AoE combos of melee DPS and Tanks, define a trackedAbility with the first
+ * abilities of the AoE and single target combos.  This is to provide leeway to finishing
+ * AoE combos when the number of targets may drop below the minimum during the combo.
+ */
+export abstract class AoEUsages extends Module {
+	static handle = 'aoeusages'
+	static title = t('core.aoeusages.title')`AoE Ability Usage`
+
+	@dependency private suggestions!: Suggestions
+
+	/**
+	 * Implementing modules MUST define the icon to be used for the suggestion.
+	 */
+	abstract suggestionIcon: string
+	/**
+	 * Implementing modules MUST define the abilities that are to be monitored.
+	 */
+	abstract trackedAbilities: AoeAbility[]
+
+	/**
+	 * Implementing modules MAY override the severity tiers for incorrect AoE usages.
+	 */
+	protected severity: SeverityTiers = {
+		1: SEVERITY.MINOR,
+		4: SEVERITY.MEDIUM,
+		7: SEVERITY.MAJOR,
+	}
+
+	private badUsages: Map<number, number> = new Map<number, number>()
```

Comment:
nit; ts is pretty magic, you can just `private badUsages = new Map<number, number>()` and it'll type it for you

Reply:
File under: things TS/JS let you do that other languages don't.  Most other languages I'm used to don't allow type inference for things other than local variables.  Will shorten in the next go